SPDHG updates including to use the stochastic sampler

MargaretDuff opened the pull request:
## Describe your changes


-Allow SPDHG to take a sampler either from our sampler class or any class with a next(self) function
-Deprecated prob from SPDHG, taking the probabilities instead from the sampler class or from a new argument prob_weights, choosing the default [1/num_subsets]*num_subsets if one is not provided in either place.
-Created two setters for the step sizes. set_step_sizes_from_ratio resets the step sizes if the user provides one/both/none of gamma and rho. step_sizes_custom takes in one/both/none of sigma and tau allowing the user to use a custom sigma and tau with those not provided calculated from the defaults. Calculating sigma from tau probably needs checking with someone else.
-Added a check_convergence function that checks self._sigma[i] * self._tau * self.norms[i]**2 <= self.prob_weights[i] for all i. This probably needs checking with someone else.
-Deprecated the kwarg "norms" to be replaced by the set_norms method in BlockOperator: added a function to return a list of norms and the ability to set this list of norms BlockOperator: added a function to return a list of norms and the ability to set this list of norms  #1513.
-Unit tests for SPDHG setters and convergence check
-Speed up of current SPDHG unit tests.

## Describe any testing you have performed
*Please add any demo scripts to [CIL-Demos/misc/](https://github.com/TomographicImaging/CIL-Demos/tree/main/misc)*
Test with SPDHG https://github.com/TomographicImaging/CIL-Demos/blob/main/misc/testing_sampling_SPDHG.ipynb

Similar results gained for all samplers for SPDHG, with 10 subsets
![image](https://github.com/TomographicImaging/CIL/assets/43645617/74937737-1e16-4d2c-95c5-37288860401d)

With 80 subsets:
![image](https://github.com/TomographicImaging/CIL/assets/43645617/e911d25a-4c17-4df6-b203-1789e2ed6bd2)




## Link relevant issues
Part of the stochastic work plan. Closes #1575. Closes  #1576. Closes #1500. Closes #1496

## Checklist when you are ready to request a review

- [x] I have performed a self-review of my code
- [x] I have added docstrings in line with the guidance in the developer guide
- [x] I have implemented unit tests that cover any new or modified functionality
- [ ] CHANGELOG.md has been updated with any functionality change
- [x] Request review from all relevant developers
- [x] Change pull request label to 'Waiting for review' 

## Contribution Notes

Please read and adhere to the [developer guide](https://tomographicimaging.github.io/CIL/nightly/developer_guide.html) and local patterns and conventions.
 - [x] The content of this Pull Request (the Contribution) is intentionally submitted for inclusion in CIL (the Work) under the terms and conditions of the [Apache-2.0](https://www.apache.org/licenses/LICENSE-2.0) License.
 - [x] I confirm that the contribution does not violate any intellectual property rights of third parties

MargaretDuff commented:
For discussion with @epapoutsellis, @jakobsj  and @gfardell  on 17/08/2023.

MargaretDuff commented:
Things that have been done:

**Stochastic Sampler**
- Implemented different sampling methods that output from a set of integers according to the method selected when either next(sampler) or sampler.next() is called
- The sampling class has a sampler.get_samples(n) method that outputs the first n samples. This does not effect/interrupt the behaviour of next
- Unit tests for the sampler 

**SPDHG**
- Allow SPDHG to take a sampler class
- Deprecate prob from SPDHG, taking the probabilities instead from the sampler class, choosing the default [1/num_subsets]*num_subsets if one is not provided in the sampler (e.g. in sequential sampling)
- Created two setters for the step sizes. set_step_sizes_from_ratio resets the step sizes if the user provides one/both/none of gamma and rho. step_sizes_custom takes in one/both/none of sigma and tau allowing the user to use a custom sigma and tau with those not provided calculated from the defaults. Calculating sigma from tau probably needs checking with someone else. 
- Added a check_convergence function that checks `self._sigma[i] * self._tau * self.norms[i]**2 <= self.prob_weights[i]` for all `i`. This probably needs checking with someone else. 
-  Deprecated the kwarg "norms" to replace with an optional argument "precalculated_norms" because I think this makes it clearer to the user what they are used for.

MargaretDuff commented:
@paskino @gfardell - A question to start discussion on this pull request. Where should the sampler class be placed in CIL - currently it is sitting in the framework folder as a separate file. It could go in the framework file, with partitioner, or it could go in the optimisation folders?

paskino commented:
It makes sense to me that this functionality goes in the `optimisation` module as it's there that it'll be most useful.

MargaretDuff commented:
Thanks @paskino, have moved it to the algorithms folder

paskino commented on docs/docs_environment.yml:
```diff
@@ -17,7 +17,7 @@
 # CIL Developers, listed at: https://github.com/TomographicImaging/CIL/blob/master/NOTICE.txt
 
-name: docs
+name: cil_testing
```
should this be here?

MargaretDuff replied:
Thanks @paskino I have added it to my gitignore about 5 times! Any tips?

paskino commented on Wrappers/Python/cil/framework/__init__.py:
```diff
@@ -34,3 +34,4 @@
 from .BlockGeometry import BlockGeometry
 from .framework import DataOrder
 from .framework import Partitioner
+from .sampler import Sampler 
```
I'm wondering where this should live. In my [comment ](https://github.com/TomographicImaging/CIL/pull/1498#issuecomment-1757553007) I suggested that it should live in the `cil.optimisation` module. 

Another option may be in `cil.utilities`.

MargaretDuff replied:
Have moved to cil.optimisation.algorithms for the moment

MargaretDuff replied:
After a discussion with Jakob and Vaggelis have moved it to cil.optimisation.utilities

paskino commented on Wrappers/Python/cil/optimisation/algorithms/SPDHG.py:
```diff
@@ -124,100 +247,118 @@ def set_up(self, f, g, operator, tau=None, sigma=None, \
+                self.prob_weights = sampler.prob
+
+        # might not want to do this until it is called (if computationally expensive)
+        self.set_step_sizes_custom()
```
This seems to be used to set the default step sizes. Why do we call this `set_step_sizes_custom`? It's confusing.

MargaretDuff replied:
Thanks edo - I added a new function set_step_sizes_default that calls set_step_sizes_custom but with no arguments passed, which sets the default step sizes

paskino commented on Wrappers/Python/cil/optimisation/algorithms/SPDHG.py:
```diff
@@ -124,100 +247,118 @@ def set_up(self, f, g, operator, tau=None, sigma=None, \
         # Choose subset
-        i = int(np.random.choice(len(self.sigma), 1, p=self.prob))
-        
+        i = self.sampler.next()
```
Or

```suggestion
        i = next(self.sampler)
```

I don't know which is better

paskino commented on Wrappers/Python/cil/optimisation/utilities/__init__.py:
```diff
@@ -1,6 +1,6 @@
 # -*- coding: utf-8 -*-
-#  Copyright 2021 United Kingdom Research and Innovation
-#  Copyright 2021 The University of Manchester
+#  Copyright 2018 United Kingdom Research and Innovation
+#  Copyright 2018 The University of Manchester
```
```suggestion
#  Copyright 2023 United Kingdom Research and Innovation
#  Copyright 2023 The University of Manchester
```

paskino reviewed (changes requested):
I think it's largely ok, but there are things to discuss/change in SPDHG

MargaretDuff commented on Wrappers/Python/cil/optimisation/algorithms/SPDHG.py:
```diff
@@ -46,15 +49,19 @@ class SPDHG(Algorithm):
-
-    **kwargs:
-    norms : list of floats
+    sampler: instance of the Sampler class
```
Thanks Edo

MargaretDuff commented on Wrappers/Python/cil/optimisation/algorithms/SPDHG.py:
```diff
@@ -63,52 +70,168 @@ class SPDHG(Algorithm):
 
         super(SPDHG, self).__init__(**kwargs)
 
+        self.prob_weights = kwargs.get('prob', None)
+        if kwargs.get('norms', None) is not None:
+            warnings.warn(
+                'norms is being deprecated, pass instead precalculated_norms=your_custom_norms')
+            if precalculated_norms is None:
+                precalculated_norms = kwargs.get('norms', None)
+
+        if self.prob_weights is not None:
+            warnings.warn('prob is being deprecated to be replaced with a sampler class. To randomly sample with replacement use "sampler=Sampler.randomWithReplacement(number_of_subsets,  prob=prob)".\
+                          If you have passed both prob and a sampler then prob will be')
```
I think I made these changes. I also agree that prob_weights is a bit tricky. We could have a chat about this?

MargaretDuff commented on Wrappers/Python/cil/optimisation/algorithms/SPDHG.py:
```diff
@@ -63,52 +70,168 @@ class SPDHG(Algorithm):
+
+    def set_step_sizes_custom(self, sigma=None, tau=None):
+        """ Sets sigma step-sizes for the SPDHG algorithm. The step sizes can be either scalar or array-objects.
+
```
Thanks - have tried to make this clearer

MargaretDuff commented on Wrappers/Python/cil/optimisation/algorithms/SPDHG.py:
```diff
@@ -63,52 +70,168 @@ class SPDHG(Algorithm):
+
+    def set_step_sizes_from_ratio(self, gamma=1., rho=.99):
+        """ Sets gamma, the step-size ratio for the SPDHG algorithm. Currently gamma takes a scalar value.
+
```
I wrote out the maths (hopefully)

MargaretDuff commented on Wrappers/Python/cil/optimisation/algorithms/SPDHG.py:
```diff
@@ -63,52 +70,168 @@ class SPDHG(Algorithm):
+    def check_convergence(self):
+        # TODO: check this with someone else
+        """  Check whether convergence criterion for SPDHG is satisfied with scalar values of tau and sigma
+
```
Done

MargaretDuff commented on Wrappers/Python/cil/optimisation/algorithms/sampler.py:
```diff
@@ -0,0 +1,461 @@
+
+    Parameters
+    ----------
+    num_subsets: int
```
Done

MargaretDuff commented on Wrappers/Python/cil/optimisation/algorithms/sampler.py:
```diff
@@ -0,0 +1,461 @@
+            The sampler will select from a list of integers {0, 1, …, S-1} with S=num_subsets. 
+
+        prob: list of floats of length num_subsets that sum to 1. default=None
+            This is the probability for each integer to be called by next. If None, then the integers will be sampled uniformly. 
```
I thought about checking this, but in reality np.random.choice will check this and it will fail there.

MargaretDuff commented on Wrappers/Python/cil/optimisation/algorithms/sampler.py:
```diff
@@ -0,0 +1,461 @@
+            Random seed for the  random number generator.  If set to None, the seed will be set using the current time. 
+
+        shuffle:boolean, default=True
+            If True, there is a random shuffle after all the integers have been seen once, if false the same random order each time the data is sampled is used.
```
Done

MargaretDuff commented on docs/docs_environment.yml:
It shouldn't be - having trouble with this file....don't want it in the pull request at all

MargaretDuff commented on Wrappers/Python/cil/optimisation/algorithms/sampler.py:
```diff
@@ -0,0 +1,461 @@
+    def next(self):
+        """ A function of the sampler that selects from a list of integers {0, 1, …, S-1}, with S=num_subsets, the next sample according to the type of sampling. """
+
+        return (self.iterator())
```
self.iterator is always configured because my methods either all hold an order in memory or call randomly with replacement based on a probability. Would there be any sampler that wouldn't do either of these?

MargaretDuff replied:
Perhaps we should have a think about the custom sampler for mantid imaging and this would help consider this class from a user's perspective?

MargaretDuff commented:
Many thanks @lauramurgatroyd for fixing the docs_build!

MargaretDuff commented on Wrappers/Python/cil/optimisation/algorithms/SPDHG.py:
```diff
@@ -124,100 +290,90 @@ def set_up(self, f, g, operator, tau=None, sigma=None, \
-
-        # initialize primal variable 
+        self.prob_weights=sampler.prob_weights #TODO: consider the case it is uniform and not saving the array 
+        if self.prob_weights is None:
```
Add this as an optional argument in the customOrder sampler and leave it out of SPDHG

MargaretDuff commented on Wrappers/Python/cil/optimisation/utilities/sampler.py:
```diff
@@ -0,0 +1,446 @@
+        return sampler
+
+    @staticmethod
+    def hermanMeyer(num_indices):
```
Make these underscores `snake` case

MargaretDuff commented on Wrappers/Python/cil/optimisation/algorithms/SPDHG.py:
```diff
@@ -46,69 +50,238 @@ class SPDHG(Algorithm):
-    def __init__(self, f=None, g=None, operator=None, tau=None, sigma=None,
-                 initial=None, prob=None, gamma=1.,**kwargs):
+
+    def __init__(self, f=None, g=None, operator=None, sigma=None, tau=None,
```
DOn't change order of sigma and tau

MargaretDuff commented on Wrappers/Python/cil/optimisation/algorithms/SPDHG.py:
```diff
@@ -46,69 +50,238 @@ class SPDHG(Algorithm):
 
+        if kwargs.get('norms', None) is not None:
+            operator.set_norms(kwargs.get('norms'))
+            warnings.warn(
```
logging not warning

MargaretDuff commented on Wrappers/Python/cil/optimisation/algorithms/SPDHG.py:
```diff
@@ -46,69 +50,238 @@ class SPDHG(Algorithm):
 
         super(SPDHG, self).__init__(**kwargs)
 
+        if kwargs.get('norms', None) is not None:
```
Move dealing with deprecated stuff to a separate function . Then move everything to a set_up method - init passes everything to set up.

MargaretDuff commented on Wrappers/Python/cil/optimisation/algorithms/SPDHG.py:
```diff
@@ -46,69 +50,238 @@ class SPDHG(Algorithm):
+    def __init__(self, f=None, g=None, operator=None, sigma=None, tau=None,
+                 initial=None, sampler=None,  **kwargs):
 
         super(SPDHG, self).__init__(**kwargs)
```
Pull out all kwarg arguments to pass to algorithm. Then check the rest of them and give an error if there is any left

MargaretDuff commented on Wrappers/Python/cil/optimisation/algorithms/SPDHG.py:
```diff
@@ -46,69 +50,238 @@ class SPDHG(Algorithm):
+                    "The value of tau should be a Number")
+            self._tau = tau
+
+    def set_step_sizes_default(self):
```
Remove this

MargaretDuff replied:
Done

MargaretDuff commented on Wrappers/Python/cil/optimisation/algorithms/SPDHG.py:
```diff
@@ -46,69 +50,238 @@ class SPDHG(Algorithm):
+        self._tau = min([value for value in values if value > 1e-8])
+        self._tau *= (rho / gamma)
+
+    def set_step_sizes_custom(self, sigma=None, tau=None):
```
Change to "set_step_sizes"

MargaretDuff commented on Wrappers/Python/cil/optimisation/algorithms/SPDHG.py:
```diff
@@ -124,100 +297,88 @@ def set_up(self, f, g, operator, tau=None, sigma=None, \
-            
-        self.g.proximal(self.x_tmp, self.tau, out=self.x)
-        
+        self.x.sapyb(1., self.zbar,  -self._tau, out=self.x_tmp)
```
Use the properties self.tau not self._tau. Similar for self._sigma

gfardell reviewed (changes requested):
It's looking good. I'm happy with the structure, just some coding comments and docstrings to address.

MargaretDuff commented on Wrappers/Python/cil/optimisation/utilities/sampler.py:
```diff
@@ -0,0 +1,679 @@
+    def next(self):
+        """ 
+
+        A function of the sampler that selects from a list of indices {0, 1, …, S-1}, with S=num_indices, the next sample according to the type of sampling.
+
```
Done

MargaretDuff commented on Wrappers/Python/cil/optimisation/utilities/sampler.py:
```diff
@@ -0,0 +1,679 @@
+        if prob_weights is None:
+            temp_list = []
+            for i in range(num_indices):
+                temp_list.append(custom_list.count(i))
```
The idea is that if I have [1,1,6,3,5] and the num_indices=8 it will output [0,2,0,1,0,1,1,0] i.e. 0 zeros, 2 ones, 0 twos, 1 three, 0 fours, 1 five, 1 six and 0 sevens? Does that make sense? Where did [2,2,1,1,1] come from?

gfardell replied:
Ah of course :D I was thinking you were using the custom list value as i.

MargaretDuff commented on Wrappers/Python/cil/optimisation/utilities/sampler.py:
```diff
@@ -0,0 +1,679 @@
+
+    def __init__(self, num_indices, order, sampling_type,  prob_weights=None):
+        """
+        The user should call this through cil.optimisation.utilities.sampler and choose the desired static method from the Sampler class.  
+        A class to select from a list of indices {0, 1, …, S-1}
+        The function next() outputs a single next index from the list {0,1,…,S-1} . Different orders are possible including with and without replacement. To be run again and again, depending on how many iterations.
```
"Returns the first `num_samples` produced by the sampler as a numpy array." Does this more sense?

MargaretDuff commented on Wrappers/Python/cil/optimisation/algorithms/SPDHG.py:
```diff
@@ -46,69 +50,255 @@ class SPDHG(Algorithm):
-                 initial=None, prob=None, gamma=1.,**kwargs):
+                 initial=None, sampler=None,  **kwargs):
+
+        max_iteration = kwargs.pop('max_iteration', 0)
+        update_objective_interval = kwargs.pop('update_objective_interval', 1)
+        log_file = kwargs.pop('log_file', None)
```
Nope...sorry found 5

MargaretDuff commented on Wrappers/Python/cil/optimisation/algorithms/SPDHG.py:
```diff
@@ -124,100 +314,91 @@ def set_up(self, f, g, operator, tau=None, sigma=None, \
-        # initialize primal variable 
+        self.ndual_subsets = self.operator.shape[0]
+        self.sampler = sampler
+        self._deprecated_kwargs(deprecated_kwargs)
```
I think this is fine. self._deprecated_kwargs uses the sampler passed to the setup (which could be None). If deprecated `prob`` is passed than self._deprecated_kwargs will set up a sampler and if not then line below will set up a sampler. It has been designed so that the function and the call can be deleted once we deprecate the arguments. 
```python
self._sampler = sampler
self._deprecated_kwargs(deprecated_kwargs)
self._sampler is None:
    self._sampler = Sampler.random_with_replacement(len(operator))
```

paskino commented on Wrappers/Python/cil/optimisation/operators/BlockOperator.py:
```diff
@@ -327,7 +327,7 @@ def __rmul__(self, scalar):
         Parameters
         ------------
-
+        
```
```suggestion
```

paskino commented on Wrappers/Python/cil/optimisation/utilities/sampler.py:
```diff
@@ -0,0 +1,764 @@
+class Sampler():
+
+    r"""
+    A class to select from a list of indices {0, 1, …, S-1}
+    The function next() outputs a single next index from the list {0,1,…,S-1} . Different orders are possible including with and without replacement. To be run again and again, depending on how many iterations.
+
+
+    Parameters
+    ----------
+    num_indices: int
+        The sampler will select from a list of indices {0, 1, …, S-1} with S=num_indices. 
+
+    sampling_type:str
+        The sampling type used. Choose from "from_function", "sequential", "custom_order", "herman_meyer", "staggered", "random_with_replacement", "random_without_replacement" and "from_function".
+
+    order: list of indices
+        The list of indices the method selects from using next. 
+
+    prob: list of floats of length num_indices that sum to 1. 
+        For random sampling with replacement, this is the probability for each index to be called by next. 
+
+    seed:int, default=None
+        Random seed for the methods that use a numpy random number generator.  If set to None, the seed will be set using the current time.  
+
```
This class follows the factory design pattern. It is not instatiated but has 7 static methods that will return instances of 7 different samplers, which require a variety of parameters. The idea of the factory is to simplify the creation of these instances with the static methods.

The beginning of the docstring does not reflect this. The examples, on the other hand do.

MargaretDuff replied:
Yes- have changed this

paskino reviewed (changes requested):
see comments

MargaretDuff commented on Wrappers/Python/cil/optimisation/algorithms/SPDHG.py:
```diff
@@ -46,69 +50,255 @@ class SPDHG(Algorithm):
+        gamma = 1.
+        rho = .99
+        if sigma is not None:
+            if len(sigma) == self.ndual_subsets:
```
It does - have added a unit test comparing it with PDHG

MargaretDuff commented on Wrappers/Python/cil/optimisation/utilities/sampler.py:
```diff
@@ -0,0 +1,764 @@
+    [6 2 1 0 4 3 5 1 0 4 2 5 6 3 3 2]
+    """
+
+    def __init__(self, num_indices, replace,  sampling_type,  prob=None, seed=None):
```
Yep, it isn't currently used, except in a unit test to check the right sampler was used. Happy to remove it? thoughts?

MargaretDuff replied:
Added a print function

MargaretDuff commented on Wrappers/Python/test/test_algorithms.py:
```diff
@@ -1079,10 +1146,11 @@ def test_SPDHG_vs_PDHG_explicit(self):
                                        1.68590e-05, decimal=3)
 
-    @unittest.skipUnless(has_astra, "ccpi-astra not available")
+"""     @unittest.skipUnless(has_astra, "ccpi-astra not available")
```
Done

MargaretDuff commented on Wrappers/Python/cil/optimisation/algorithms/SPDHG.py:
```diff
@@ -120,16 +150,18 @@ def __init__(self, f=None, g=None, operator=None, tau=None, sigma=None,
         return_all=kwargs.pop('return_all', False)
```
Done

MargaretDuff commented on Wrappers/Python/cil/optimisation/operators/BlockOperator.py:
Done

MargaretDuff commented:
Updated to include just updates to SPDHG:
- Allow SPDHG to take a sampler either from our sampler class or any class with a __next__(self) function 
- Deprecated prob from SPDHG, taking the probabilities instead from the sampler class or from a new argument `prob_weights`, choosing the default [1/num_subsets]*num_subsets if one is not provided in either place. 
-Created two setters for the step sizes. set_step_sizes_from_ratio resets the step sizes if the user provides one/both/none of gamma and rho. step_sizes_custom takes in one/both/none of sigma and tau allowing the user to use a custom sigma and tau with those not provided calculated from the defaults. Calculating sigma from tau probably needs checking with someone else.
- Added a check_convergence function that checks self._sigma[i] * self._tau * self.norms[i]**2 <= self.prob_weights[i] for all i. This probably needs checking with someone else.
-Deprecated the kwarg "norms" to be replaced by the set_norms method in BlockOperator: added a function to return a list of norms and the ability to set this list of norms  #1513.
-Unit tests for SPDHG setters and convergence check
-Speed up of current SPDHG unit tests.
-

MargaretDuff commented:
Replaced by #1644 and #1643
